### PR TITLE
[no ticket][risk=no] Puppeteer refactor findOrCreateWorkspace function

### PR DESCRIPTION
**Pull request by aweng98**
Fix test failures caused by data concurrency problem. happens when same workspace is used by more than one tests running concurrently.

`findOrCreateWorkspace` function returns workspaces older than 30 minutes or create new if none found.

### CODE REVIEW

File: e2e/app/component/workspace-card.ts [hunk 1]
```diff
@@ -35,20 +36,31 @@ export default class WorkspaceCard extends CardBase {
...
       await page.waitForXPath(WorkspaceCardSelector.cardRootXpath, {visible: true, timeout: 1000});
     } catch (e) {
       return [];
     }
-    const cards = await page.$x(WorkspaceCardSelector.cardRootXpath);
+    const cardElements = await page.$x(WorkspaceCardSelector.cardRootXpath);
     // transform to WorkspaceCard object
-    const resourceCards = cards.map(card => new WorkspaceCard(page).asCard(card));
-    return resourceCards;
+    const workspaceCards = cardElements.map(card => new WorkspaceCard(page).asCard(card));
```
Comment (ericsong) [inline e2e/app/component/workspace-card.ts:51]: nit: u can get rid of the `cardElements` variable and just combine the statements

File: e2e/app/component/workspace-card.ts [hunk 2]
```diff
@@ -35,20 +36,31 @@ export default class WorkspaceCard extends CardBase {
...
+
+    const filtered = [];
+    if (accessLevel !== undefined) {
+      for (const card of workspaceCards) {
+        const cardAccessLevel = await card.getWorkspaceAccessLevel();
+        if (cardAccessLevel === accessLevel) {
+          filtered.push(card);
+        }
+      }
+      return filtered;
```
Comment (ericsong) [inline e2e/app/component/workspace-card.ts:61]: actually, you can combine this too.

`return (await page.$x(WorkspaceCardSelector.cardRootXpath)
                .map(card => newWorkspaceCard(page).asCard(card)
                .filter(card => !accessLevel || (await card.getWorkspaceAccessLevel()) === accessLevel)`

File: e2e/utils/test-utils.ts [hunk 3]
```diff
@@ -224,23 +224,36 @@ export async function findOrCreateWorkspace(page: Page, opts: {alwaysCreate?: bo
...
+  // Find all workspaces with OWNER role
+  const existingCards = await WorkspaceCard.findAllCards(page, WorkspaceAccessLevel.Owner);
 
-  const existingWorkspaces = await workspaceCard.getWorkspaceMatchAccessLevel(WorkspaceAccessLevel.Owner);
-  if (alwaysCreate || existingWorkspaces.length === 0) {
-    // Create new workspace
+  // Filter to include Workspaces older than 30 minutes
+  const halfHourMillisec = 1000 * 60 * 30;
+  const now = Date.now();
+  const filtered = [];
```
Comment (ericsong) [inline e2e/utils/test-utils.ts:233]: I think you can do a similar refactoring here with `existingCards` and `filtered`. Create a variable like... `const olderWorkspaceCards = findAllCards().filter()`

Aside from cutting down on some lines of code, it reduces the number of variables declared. Also, a variable name like `olderWorkspaceCards` provides a little bit on context of how the cards were filtered.

File: e2e/app/component/workspace-card.ts (hunk 2 above)
Comment (aweng98) [reply e2e/app/component/workspace-card.ts:60]: Good suggestion. I combined some code. But unable to use  `.filter` out of the box with async function . Googled for how to, I think I'll needs new custom code to make filter work with async function.

**Review by ericsong — APPROVED** [review]

**Pull request merged**